README
How to use 'Mave' textbase mailer.

Requirement

  Ruby, curses and qdbm/gdbm library.

1. Get archive and Extract

  Access http://sourceforge.jp/projects/mave/downloads/xxxxx/mave-2.98.tgz/
  $ tar xvfz mave-2.98.tgz
  $ cd mave-2.98

2. Build your config

  $ cp mave.config.sample mave.config
  $ vi mave.config

    rewrite setting of 'Main' account to yours.
    You can jump to paragraph 3, if you try mave immediately.

    configure your editor environment.

    @configs[:EDITOR_TYPE] = ['system'|'forkexec']
        system: Mave close curses window and wake editor with same console. For CUI.
      forkexec: Mave fork child process and wake editor another window. For GUI.

    @configs[:EDITOR] = '/usr/bin/editor %s'
      The command of wake editor. %s is replaced to file name.

    configure your viewer environment.

    @configs[:VIEWER_TYPE] = ['system'|'forkexec']
    @configs[:VIEWER] = '/usr/bin/viewer %s'
      These settings are similar to editor's.

    configure your file management environment.

    @configs[:FILE_MANAGER_TYPE] = ['system'|'forkexec']
    @configs[:FILE_MANAGER] = '/usr/bin/nautilus %s >/dev/null 2>&1'
      These settings are used for open directory by file manager for manipulate files.

    configure file identification command.

    @configs[:FILE_IDENTIFIER] = '/usr/bin/file -b -i %s 2>/dev/null'
      These settings are used for identify file types when attach file to mail and etc.

3. Make directory for store mail files

  $ mkdir ~/mave.mails

  Mave do not create any files without the directory specified here.

  To send mail, you should build the address book this phase.
  See the paragraph of "15. Build your addressbook" if you want.

4. Execute Mave

  $ ./mave

  If screen is garbled, see the paragraph of "Note:".

5. Fetch mails

  Press Ctrl + x, p key to fetch mails from server.

  In default, Mave does not remove mails on pop server immediately.
  The time lag until removing mails was set by ':POP_KEEP_TIME'.
  It is possible to try use Mave in parallel with your current mailer.

  Status window is scrolled by Pgup, PgDn keys.

6. Explore mails

  Each h, j, k, and l key were allocated in each Left, Down, Up, and Right like the Vi editor.

  How to look summary line.

  x.  22 Amazon.com   <sh Sun, 22 Nov 15:40  579 .. +Amazon.com order confirmation
  ^^ ^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^^ ^^^^ ^^ ^ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ||  |   Mail from / to   Sent date        size || |  Subject
  ||  +-Recent 0 <=> 999 Old                     || +-Parent(+) / Child(-) / Folded(#/=)
  |+-Flag(F) / Notice mark(#)                    |+-Multipart mail(@)(Has attachment)
  +-Unread mark(x)                               +-To you(w) / Cc you(v)

  You can change summary looking.

  r key: Unread mark ON/OFF           f key: Flag ON/OFF
  o key: Fold thread / Open           n key: Notice mark ON/OFF

  Press Ctrl + x, v key to open mail viewer.

7. Incremental search

  Press Ctrl + s key to open incremental search dialog at summary or preview view.
  The search will running that each key was pressed.
  Press Ctrl + s key to search next when word found.
  Press Ctrl + g key to abort search or close search dialog.
  Press Ctrl + r key to search backward.

  You can incremental search with romanized japanese key stroke at summary view,
  by installing the kanji kana simple inverter KAKASI and its ruby bindings.
  This function is build by convert to romanized japanese and its caching.
  Press ESC, Shift + q key to see the cache data.

8. Create/Edit mail message

  Press Ctrl + x, n key to create new mail message.
  Press Ctrl + x, r key to create reply mail message.
  The message will store to 'Drafts' folder.

  Press Ctrl + x, e key to edit mail message.
  The message will overwrite.

  At the headers of destination (To:, Cc:, Bcc:), you can specify the receiver by entries on address book.
  The word enclosed with two hyphens is recognized as ID that defined in the address book, and converted as follows.

  Example of definition)
    entries << { :MAIL => 'david-x@google.com',                     :ID => 'david',                                 }
    entries << { :MAIL => 'john-y@yahoo.co.jp',                     :ID => 'john@private',  :SEND => 'johnny',      }

  At Edit)                     At Send)
    To: --david--,          =>   To: david <david-x@google.com>, johnny <john-y@yahoo.co.jp>
        --john@private--

9. Mail Folders

  Press h key on the summary view, change into the folder view.
  In folder view, select folder with up/down(k, j), and move into the folder with right(l).
  Or press shortcut key to move into the folder.

  How to look folder list.

  x 0) Inbox (5/52)
  . a) Community (2/38)
  x b) Develop (1/8)
  . d) Drafts (0/23)
  . o) Outbox (0/0)
  . s) Sent (0/25)
  . t) Trash (0/18)
  ^ ^  ^^^^^  ^^^^
  | |    |    Number of unread / all mails
  | |    +-Folder name
  | +-Shortcut key
  +-Folder got new mails mark(x)

  Press r key to ON/OFF got new mails mark.
  In folder view, press Ctrl + x, n key to create new folder.
  In selecting folder, press Ctrl + x, e key to edit folder configurations.
  You can change display name, list order and mail distributing by edit folder configurations.
  See example in 'mave.folderconfig.master'.

10. Move, Copy and Delete mails.

  In summary view, press Ctrl + x, m key to move mail to another folder.
  The folder view opens, and select the destination folder.

  Press Ctrl + x, c key to copy mail to another folder.

  Press Ctrl + x, d key to delete mail.
  The mail deleted from except 'Trash' folder, it moves to the 'Trash' folder.
  The mail deleted from 'Trash' folder, it deleted completely.

11. Send mails

  Press Ctrl + x, s key to send mails in 'Outbox' folder.

  Usually, the edited E-mail is stored in 'Draft' folder,
  and moves to 'Outbox' folder to send when mail is finished writing.

  Sent mails are moved to 'Sent' folder.

12. Import mails

  Press Ctrl + x, Shift + i key to import mails according to the setting of each account.

  Import mails are shown by the command that lists the import file with a standard output.

  Example)
    account[:IMPORT_COMMAND] = %Q!/usr/bin/find /home/old_user/mave.mails -name '*.eml' |
                                  grep -E '/Inbox/' | sed 's/\\/.*\\//& /'| sort -k 2 | sed 's/ //'!

  Or, it describes to the text file, and execute cat command for it.
  There should be import mails in the format of one mail by one file.
  Now, Mave doesn't correspond to the mbox format and other.

  The default config has the setting that import manual of Mave that formatted like mail.
  Press Ctrl + x, Shift + i key to import its when just start Mave.

13. Mark mails

  In summary view, press m key to mark mail.
  The marked mails are indicated with 'M'.

  - x.  22 Amazon.com   <sh Sun, 22 Nov 15:40  579 .. +Amazon.com order confirmation
  M x.  22 Amazon.com   <sh Mon, 23 Nov 15:41  612 ..   -Amazon.com shipping information (#123-1423534-1048639)
  ^ "M" mark

  Press u key to unmark mail.
  Press ESC, u key to unmark all mails.

  Most functions act on marked mails while marking mails.

14. Arrange mail threads

  In summary view, Press Ctrl + x, Shift + j key to take off mail from thread.

  Press ESC, j key to tie mail to original thread.

  Press Ctrl + x, j key to tie mail to mail thread that you want to.
  The mail that marked will tie to the mail that indicated with cursor, when press Ctrl + x, j key.

15. Build your addressbook

  $ cp mave_address.sample mave_address
  $ vi mave_address

  Fill forms with family's, friend's and your mail address, according to the sample.

  Example)
    entries << { :MAIL => 'david-x@google.com',                     :ID => 'david',                                 }
    entries << { :MAIL => 'john-y@yahoo.co.jp',                     :ID => 'john@private',  :SEND => 'johnny',      }
    entries << { :MAIL => 'linus-z@abc.co.jp',                      :ID => 'linus@ABC',     :SEND => 'Mr. Zorvalds',}

  Entries were registered to the address book DB by the execute script of address book.

  $ ./mave_address

16. Attach file to mail

  To viewing attachment files in received mail.

    In summary view, Press Ctrl + x, x key to extract all attachment files in mail to related directory.
    The related directory is paired for each mail, that holds some files.
    In summary view, Press ESC, f key to open related directory by file manager.

  To replying mail with attachment files that had some modify.

    In summary view, Press Ctrl + x, r key to create reply mail that has X-Mave-Extract-Targets header for file extract.
    After edit and save mail, attachment files are extract to related directory described under X-Mave-Extract-Targets header.
    Press ESC, f key for reply mail to open related directory by file manager and modify it.
    Press Ctrl + x, e key to edit mail. Extracted files are described under X-Mave-Relations header.
    You can move file description to under X-Mave-Attachments header from under X-Mave-Relations header, if you want to attach and reply.
    Move mail to 'Outbox' folder and press Ctrl + x, s key to send mail with attachment files.

  To send new mail with new attachment files.

    In summary view, Press Ctrl + x, n key to create new mail that has empty X-Mave-Relations header.
    Describe filenames that you will attach under X-Mave-Relations header, empty files that have same filenames will create in related directory.
    Press ESC, f key for new mail to open related directory by file manager and modify it.
    You can send mail with attachment files by move file description to under X-Mave-Attachments header from under X-Mave-Relations header.

17. Clip(quote) and Tagjump

  Clip(quote) function is a function that quotes a part of mail, and pastes it on another mail.
  The quotes has a message-id(tag) in mail. It is possible to jump(tagjump) to original mail.
  If these functions are used, the schedule management can be done on mailer, and the full-text search capability described later be achieved.

  In preview view, the "]" mark where the several lines from top in the left is a cursor. It can used for the clip from mail.
  Target cursor to the line that wants to begin quoting, and press Ctrl + c to clip line to clip file.
  The cursor moves to the next line after quoting, you can press more Ctrl + c to clip multi lines continuously.

  In default, the clip file will create with mave.clip file under _pop under the mail directory.
  You can paste this file, by using the file insertion function on the ediror.

  The quote lines are the following forms.

      * [MEETING] Mailer Mave 2.97 last review
        <1268959354.007885.8ed5d181@mave.zakato.itline.jp> line:18
      - Thursday, April 1, at 1:15 p.m to 3:00 p.m
      - Meeting room 503

  Target cursor to message-id tag on second line of quotes, and press Crtl + ] to jump to original mail.
  The position that before the tag jump is store to the tagstack, you can return immediately former positon by press Ctrl + t.

18. Full-text search

  You can the full-text search of the content of mail each folder, by installing the full-text search engine.
  The full-text search function is build by combining shell command execution, import result and tagjump functions.
  The interface for Hyper Estraier and Namazu is prepared as a scripting example.
  See the fulltext_search method in mave.folderconfig.common for the example of scripting the interface.

  Move to the target folder that wants to do the full-text search, and press ESC, s at summary or preview view to open the full-text search dialog.
  Input word or phrased for search and press Enter to execute search and result is stored to the folder as a mail format.
  You can jump to original mail by using the tagjump function on the search result.

19. Execute shell command and import result

  You can execute shell command that described in in mail, and result is stored to the folder as a mail format.

  Target cursor to line that has shell command, and press ESC, ! to execute it.
  If the execution result pseudotakes the form of mail, the mail header information is effect.
  Subject and Message-ID of the taken mail can be controlled by using this.
  And the store folder can be specified by using X-Mave-Store-Folder pseudoheader.
  See imported mail (Subject: example of the script launcher) for the example of scripting the interface.

20. Prevent to forget to attach to mail. Pre-sent and post-send script.

  You can execute the script when just before and after send mail.
  This function used for prevent to forgot to attach to mail by word checking in mail body when just before send mail.
  See the presend method in mave.folderconfig.common for the example of scripting.

21. Quit

  Press Ctrl + x, c key to kill mave.

22. Web mail viewer 'Maverick'
  
  The 'Maverick' is cgi script of mail viewer for mail reading with web browser.
  Mail data and index of full-text search are completely shared with mave.
  To use, install Apache and eRuby, move mave directory under public_html.
  Don't forget to set access permittion properly refer dot.htaccess.
  You can use like Gmail by cyclic execute mave_fetch script that attached.

Enjoy!!

Note:
  When multibyte character is troubled.
  Standard curses library of Ruby doesn't correspond to UTF-8.
  Please apply the patch to correspond to ncursesw referring to the following sites.

  http://www.itline.jp/~svx/diary/?date=20091022

  In experimentally, Mave corresponds terminal environment of EUC-JP now.
  In this case, you do not need to apply above patch.
  Set EUC-JP to ':TERMINAL_CHARSET' in mave.config, and change mave_base.rb a bit.
  "export LANG=ja_JP.eucjp" might execute from the terminal.

  117     def enspc
  118 #       return(self)                                            # comment out without UTF-8
  119 ^       self.gsub(/[\xC0-\xE2][\x80-\xBF]+/) {|c|               #### for UTF8 temporally
      +-remove this '#'

  And in experimentally, Mave is corresponding Windows console environment with SHIFT_JIS.
  Set SHIFT_JIS to ':TERMINAL_CHARSET' in mave.config, and try and error.

Mave project official page:
  http://sourceforge.jp/projects/mave/

[EOF]


CHANGELOG
Changelog of Textbase Mailer 'Mave'

Version 2.99
	* Refine for ruby 2.0.
	* Add wiki edit function.

Version 2.98
	* Add web mail viewer 'Maverick'.
	* Add incremental search functions with romanized japanese stroke.
	* Add QDBM support for meta data store of mail.
	* Add legacy encoding mode for filename of attachment.
	* Add import address book function from ldap server.

Version 2.97
	* Add clip(quote) and tagjump function.
	* Add full-text search function for mail by external engine.
	* Add execute shell command and import result function.
	* Add execute present and postsend script function.(For prevent to forget to attach to mail)

Version 2.96
	* Add extract/enclose attachment files function.(Supprt RFC 2231 extensions)
	* Add ssl/tls protocol support.(Send/Receive Gmail account)
	* Correct permissions of related files.

Version 2.95
	* Add send mail function.
	* Add address book function.
	* Add filter and distribute mail function.
	* Add import mail function.
	* Add mail viewer(full screen) function.

Version 2.94
	* Add mail move, copy and delete functions.
	* Add arrange functions of mail thread relations.

Version 2.93
	* Add incremental search functions for summary or preview view.
	* Add move into the folder function by shortcut key.

Version 2.92
	* Add folder functions.

Version 2.91
	* Add message edit functions.

Version 2.90
	* Initial commit.
	* It can only fetch and explore mails.

[EOF]


mave_base.rb
# coding: utf-8

require 'kconv'

class IO
	def getc
		RUBY_VERSION < '1.9' ? super : self.read(1).ord
	end
end

#===============================================================================
#
#	Integer
#
class Integer

	#-----------------------------------------------------------
	#
	#	数値を人間に読みやすい表現形式で返す
	#
	def to_h(dot = false)
		unless((full = self.to_s).length > 3)
			full
		else
			case full.length % 3
				when 1 then top = full[0, 1] + (dot ? '.' + full[1, 1] : '')
				when 2 then top = full[0, 2]
				when 0 then top = full[0, 3]
			end
			top + '....KKKMMMGGGTTTPPP'[full.length].chr
		end
	end
end

#===============================================================================
#
#	String
#
class String

	if(RUBY_VERSION < '1.9')
		def ord
			self[0]
		end

		def force_encoding(encoding)
			self
		end
	end

	XFF = "\xFF".force_encoding('ASCII-8BIT')
	XFFFF = XFF * 2

	#-----------------------------------------------------------
	#
	#	キャラクタセットの定義
	#
	#		http://www.iana.org/assignments/character-sets
	#
	@@charsets = {
		'ISO-2022-JP'	=> Kconv::JIS,
		'SHIFT_JIS'		=> Kconv::SJIS,
		'SHIFT-JIS'		=> Kconv::SJIS,
		'EUC-JP'		=> Kconv::EUC,
		'UTF-8'			=> Kconv::UTF8,
	}

	def self.charset(charset)
		@@charsets[charset]
	end

	#-----------------------------------------------------------
	#
	#	メッセージヘッダのデコーダ
	#
	#		http://tools.ietf.org/html/rfc2047
	#
	@@decoders = {}
	@@current_decode_charset = 'UTF-8'

	def self.bind_decoder(code)
		@@decoders[code.upcase] = Proc.new
	end

	def self.set_decode_charset(charset)
		@@current_decode_charset = charset
	end

	def decode_mh												# decode message header
		gsub(/=\?([^?]+)\?(B|Q)\?([^?]+)\?=/i) {
			$3.decode_ec($2).decode_cs(@@current_decode_charset, $1)
		}.gsub(/[\x00-\x1F]/) {|c|
			'\x%02X' % c.ord
		}.encode(@@current_decode_charset, :invalid => :replace, :undef => :replace)
	end

	def decode_ec(code)											# decode encodings
		(it = @@decoders[code.upcase]) ? it.call(self) : self
	end

	def decode_cs(out_code, in_code)							# decode character sets
		(it = @@decoders[in_code.upcase]) ? it.call(self, out_code) : self.inspect
	end

	#-----------------------------------------------------------
	#
	#	メッセージヘッダのエンコーダ
	#
	#		http://tools.ietf.org/html/rfc2047
	#
	@@encoders = {}
	@@current_encode_charset = 'ISO-2022-JP'
#	@@current_encode_charset = 'UTF-8'
	@@current_encode_encoding = 'B'
#	@@current_encode_encoding = 'Q'

	def self.bind_encoder(code)
		@@encoders[code.upcase] = Proc.new
	end

	def self.set_encode_charset(charset)
		@@current_encode_charset = charset
	end
	def encode_mh_multi(field_name)								# encode message header multi line
		begin		
			(it = @@encoders[('<%s><%s><%s>' % ['MULTI', @@current_encode_charset, @@current_encode_encoding]).upcase]) ? it.call(self, field_name, Proc.new) : raise
		rescue
			field_name + ': ' + self.encode_mh
		end
	end
	def encode_mh												# encode message header
		"=?%s?%s?%s?=" % [@@current_encode_charset, @@current_encode_encoding,
			self.encode_cs(@@current_encode_charset, 'UTF-8').encode_ec(@@current_encode_encoding).chomp]
	end

	def encode_ec(code)											# encode encodings
		(it = @@encoders[code.upcase]) ? it.call(self) : self
	end

	def encode_cs(out_code, in_code)							# encode character sets
		decode_cs(out_code, in_code)
	end

	def encode_body(code)										# encode message body
		encode_cs(code, @@current_decode_charset)
	end

	#-----------------------------------------------------------
	#
	#	RFC 2231 拡張表現のエンコーダ
	#
	#		http://tools.ietf.org/html/rfc2231
	#
	@@rfc2231_encoders = {}
#	@@current_rfc2231_encode_charset = 'UTF-8'
	@@current_rfc2231_encode_charset = '<LEGACY><ISO-2022-JP><B>'

	def self.bind_rfc2231_encoder(code)
		@@rfc2231_encoders[code.upcase] = Proc.new
	end

	def self.set_rfc2231_encode_charset(charset)
		@@current_rfc2231_encode_charset = charset
	end

	def rfc2231_encode(attr, n = 78)
		@@rfc2231_encoders[@@current_rfc2231_encode_charset].call(self, attr, n, Proc.new)
	end

	#-----------------------------------------------------------
	#
	#	value のエンコーダ
	#
	def value_encode
		self =~ /[^-.0-9A-Z_]/i ? "\"#{self.gsub(/"/, '\"')}\"" : self
	end

	#-----------------------------------------------------------
	#
	#	ext-octet のデコーダ、エンコーダ
	#
	def ext_decode
		self.gsub(/%([0-9A-F]{2})/i) { $1.to_i(16).chr }
	end
	def ext_encode
		self.gsub(/[^-.0-9A-Z_]/i) {|c| '%%%02X' % c[0] }
	end

	#-----------------------------------------------------------
	#
	#	文字列を指定の長さに切り詰める
	#
	@@wsizer = {}
	@@centerer = {}
	@@snippers = {}
	@@each_snippers = {}
	@@current_snip_charset = 'UTF-8'
#	@@current_snip_charset = 'EUC-JP'
#	@@current_snip_charset = 'SHIFT_JIS'

	def self.bind_wsizer(charset)
		@@wsizer[charset] = Proc.new
	end

	def self.bind_centerer(charset)
		@@centerer[charset] = Proc.new
	end

	def self.bind_snipper(charset)
		@@snippers[charset] = Proc.new
	end

	def self.bind_each_snipper(charset)
		@@each_snippers[charset] = Proc.new
	end

	def self.set_snip_charset(charset)
		@@current_snip_charset = charset
	end

	def wsize
		@@wsizer[@@current_snip_charset].call(self)
	end

	def center(n, padding = ' ')
		@@centerer[@@current_snip_charset].call(self, n, padding)
	end

	def snip(n, charset = @@current_snip_charset)
		@@snippers[charset].call(self, n)
	end

	def each_snip(n, max = 9999)
		@@each_snippers[@@current_snip_charset].call(self, n, max, Proc.new)
	end

	#-----------------------------------------------------------
	#
	#	端末の UTF-8 対応の不備(記号の幅)を補う
	#
	#		http://ja.wikipedia.org/wiki/UTF-8
	#
	def enspc
#		return(self)											# UTF-8 以外ならコメントを生かす
		self.force_encoding('ASCII-8BIT').gsub(/[\xC0-\xE2][\x80-\xBF]+/n) {|c|	#### for UTF-8 いーかげん
			c + ' '
		}
	end

	#-----------------------------------------------------------
	#
	#	Re: をまとめる
	#
	def group_re(level = 0, re = 'Re')
		base = self.dup
		while(base =~ /^\s*#{re}\^?\d*:/i)
			base.sub!(/^\s*#{re}\^?(\d*):\s*/i) {
				level += ($1.to_i > 1 ? $1.to_i : 1)
				''
			}
		end
		(level < 1 ? '' : "#{re}: ") + base
#		(level < 1 ? '' : "#{re}#{level < 2 ? '' : "^#{level}"}: ") + base  # Re^3 表記
	end

	#-----------------------------------------------------------
	#
	#	Fw: をまとめる
	#
	def group_fw(level = 0)
		group_re(level, 'Fw')
	end

	#-----------------------------------------------------------
	#
	#	HTML の charset を返す
	#
	def html_charset
		(self =~ /"text\/html; charset=(.+)"/) ? $1 : nil
	end

	#-----------------------------------------------------------
	#
	#	HTML を UTF-8 に
	#
	def html_utf8
		(it = html_charset) ?
			self.encode('UTF-8', it, :invalid => :replace, :undef => :replace) :
			self.encode('UTF-8',     :invalid => :replace, :undef => :replace)
	end
end

#===============================================================================
#
#	多言語対応クラス
#
#		http://www.gnu.org/software/gettext/gettext.html
#
class Intl

	@@domains = {}
	@@domains[@@current_domain = 'default'] = {}

	def self.bind_text_domain(domain, dirname = '')
		load "#{dirname}#{domain}.pmo"
		@@domains[domain] = @@catalog
	end

	def self.set_text_domain(domain)
		@@current_domain = domain
	end

	def self.get_text(msgid)
		@@domains[@@current_domain][msgid] || msgid
	end
end

#===============================================================================
#
#	多言語対応
#
def _(msgid)
	Intl.get_text(msgid)
end

#===============================================================================
#
#	その他
#
def yap(arg = 'done.')
	@yap = 0 unless(@yap)
	print "#{@yap += 1}: #{arg.inspect}\n"
end

def debug(log = 'log.', obj = self)
	@debug = File.new('debug.log', 'a') and @debug.write('-' * 76 + "\n") unless(@debug)
	@debug.write(obj.to_s + ': ' + log.to_s + "\n")
end

#===============================================================================
#
#	各種デコーダ/エンコーダを登録
#
String.bind_decoder('7BIT') {|str|								# 7bit decoder
	str
}
String.bind_decoder('8BIT') {|str|								# 8bit decoder
	str
}
String.bind_decoder('BINARY') {|str, out_code|					# binary decoder?
#	'-- binary --'
	str.inspect
}
String.bind_decoder('BASE64') {|str|							# Base64 decoder
	str.unpack('m')[0]
}
String.bind_decoder('QUOTED-PRINTABLE') {|str|					# Quoted Printable decoder
	str.unpack('M')[0]
}
String.bind_decoder('B') {|str|									# Base64 decoder
	str.unpack('m')[0]
}
String.bind_encoder('B') {|str|									# Base64 encoder
	[str].pack('m999')
}
String.bind_decoder('Q') {|str|									# Quoted Printable decoder
	str.unpack('M')[0]
}
String.bind_encoder('Q') {|str|									# Quoted Printable encoder
	[str].pack('M999')
}
String.bind_encoder('<MULTI><ISO-2022-JP><B>') {|str, field_name, proc|	# encode message header multi line
	src = str + ' '; single_max = (max_length = 76) - (line = field_name + ': ').length
	while(src.length > 1)
		multi_max = ((single_max - 18) / 4 * 3 - 6) / 2
		if(single_max > 0 and src.force_encoding('ASCII-8BIT')  =~ /^([\x20-\x7E]{1,#{single_max}})[\x20\xC0-\xFD]/n)
			line += src.slice!(0, $1.length)
			single_max -= $1.length
		elsif(multi_max > 0 and src.force_encoding('ASCII-8BIT') =~ /^(([\xC0-\xFD][\x80-\xBF]+){1,#{multi_max}})/n)
			line += (line0 = src.slice!(0, $1.length).encode_mh)
			single_max -= line0.length
		else
			line == '' and raise
			proc.call(line.gsub(/^\t\s/, "\s")); line = "\t"; single_max = max_length
		end
	end
	proc.call(line.gsub(/^\t\s/, "\s"))
}
String.bind_encoder('<MULTI><UTF-8><B>') {|str, field_name, proc|	# encode message header multi line
	src = str + ' '; single_max = (max_length = 76) - (line = field_name + ': ').length
	while(src.length > 1)
		multi_max = (single_max - 12) / 4 * 3
		if(single_max > 0 and src.force_encoding('ASCII-8BIT') =~ /^([\x20-\x7E]{1,#{single_max}})[\x20\xC0-\xFD]/n)
			line += src.slice!(0, $1.length)
			single_max -= $1.length
		elsif(multi_max > 0 and src.force_encoding('ASCII-8BIT') =~ /^([\x80-\xFD]{1,#{multi_max}})[\x20-\x7E\xC0-\xFD]/n)
			line += (line0 = src.slice!(0, $1.length).encode_mh)
			single_max -= line0.length
		else
			line == '' and raise
			proc.call(line.gsub(/^\t\s/, "\s")); line = "\t"; single_max = max_length
		end
	end
	proc.call(line.gsub(/^\t\s/, "\s"))
}
String.bind_decoder('US-ASCII') {|str, out_code|				# us-ascii decoder
	str
}
String.bind_decoder('ISO-8859-1') {|str, out_code|				# iso-8859-1 decoder
	str
}
String.bind_decoder('ISO-2022-JP') {|str, out_code|				# iso-2022-jp decoder
	str.kconv(String.charset(out_code), Kconv::JIS)
}
String.bind_decoder('ISO-2022-JP-1') {|str, out_code|			# iso-2022-jp-1 decoder
	str.kconv(String.charset(out_code), Kconv::JIS)
}
String.bind_decoder('ISO-2022-JP-2') {|str, out_code|			# iso-2022-jp-2 decoder
	str.kconv(String.charset(out_code), Kconv::JIS)
}
String.bind_decoder('SHIFT_JIS') {|str, out_code|				# shift_jis decoder
	str.kconv(String.charset(out_code), Kconv::SJIS)
}
String.bind_decoder('SHIFT-JIS') {|str, out_code|				# shift_jis decoder
	str.kconv(String.charset(out_code), Kconv::SJIS)
}
String.bind_decoder('EUC-JP') {|str, out_code|					# euc-jp decoder
	str.kconv(String.charset(out_code), Kconv::EUC)
}
String.bind_decoder('UTF-8') {|str, out_code|					# utf-8 decoder
	str.kconv(String.charset(out_code), Kconv::UTF8)
}
#	http://tools.ietf.org/html/rfc2152
String.bind_decoder('UTF-7') {|str, out_code|					# utf-7 decoder
	str.gsub(%r|\+([A-Za-z0-9+/]+)-?|) {|p|
		($1 + '==').unpack('m')[0].kconv(Kconv::UTF8, Kconv::UTF16)
	}.gsub(/\+-/, '+').kconv(String.charset(out_code), Kconv::UTF8)
}

# 11bit: 0xC0-0xDF 0x80-0xBF
# 16bit: 0xE0-0xEF 0x80-0xBF 0x80-0xBF
# 21bit: 0xF0-0xF7 0x80-0xBF 0x80-0xBF 0x80-0xBF
# 26bit: 0xF8-0xFB 0x80-0xBF 0x80-0xBF 0x80-0xBF 0x80-0xBF
# 31bit: 0xFC-0xFD 0x80-0xBF 0x80-0xBF 0x80-0xBF 0x80-0xBF 0x80-0xBF

String.bind_wsizer('UTF-8') {|str|								# 表示幅を得る
	str.force_encoding('ASCII-8BIT').gsub(/[\xC0-\xFD][\x80-\xBF]+/n, "\xFF\xFF").length
}

String.bind_centerer('UTF-8') {|str, n, padding|				# センタリングする
	w = n - str.wsize
	w = 0 if(w < 0)
	(padding * (w >> 1) + str + padding * w).snip(n)
}

String.bind_snipper('UTF-8') {|str, n|							# 指定の長さに切り詰める
	str.force_encoding('ASCII-8BIT')
	ws = str[0, n * 2].gsub(/[\xC0-\xFD][\x80-\xBF]+/n, String::XFFFF)[0, n]
	wc = ws.count(String::XFF)
	str.slice(0, n + wc / 2 - wc % 2) + ' ' * (n - ws.length + wc % 2)	# ASCII-8BIT で返す
}

String.bind_each_snipper('UTF-8') {|str, n, max, proc|			# 指定の長さに切り詰め、順に行を渡す
	str.force_encoding('ASCII-8BIT')
	p = 0; while(p <= str.length)								# '<': 改行文字のみの行は省略
		break if((max -= 1) < 0)
		ws = str[p, n * 2].gsub(/[\xC0-\xFD][\x80-\xBF]+/n, String::XFFFF)[0, n]
		wc = ws.count(String::XFF)
		proc.call(str.slice(p, nn = n + wc / 2 - wc % 2) + ' ' * (n - ws.length + wc % 2))
		p += nn
	end
}

String.bind_rfc2231_encoder('UTF-8') {|str, attr, n, proc|		# RFC 2231 拡張表現にエンコードして返す(添付ファイル名指定用)
	multi = false; head = ''; str.force_encoding('ASCII-8BIT') =~ /[\xC0-\xFD]/n and multi = true and head = "utf8''"
	w = n - attr.size - "\t*n*=;".size - head.size
	nth = -1; nl = ''; el = ''; str.gsub(/./u) {|nc|
		multi = true if(!multi and nc.force_encoding('ASCII-8BIT') =~ /[\xC0-\xFD]/n)
		if(lnl = nl and lel = el and nl += nc and (el += (ec = nc.ext_encode)).size > w)
			proc.call("\t%s*%d%s=%s%s;" % [attr, nth += 1, multi ? '*' : '', head, multi ? lel : lnl.value_encode])
			multi = false; head = ''
			w = n - attr.size - "\t*n*=;".size - head.size
			nl = nc; el = ec
		end
	}
	proc.call("\t%s%s%s=%s%s" % [attr, nth == -1 ? '' : "*#{(nth += 1).to_s}", multi ? '*': '', head, multi ? el : nl.value_encode])
}

String.bind_rfc2231_encoder('<LEGACY><ISO-2022-JP><B>') {|str, attr, n, proc|	# RFC 違反だが B encoding でエンコードして返す(添付ファイル名指定用)
	proc.call("\t%s=\"=?%s?%s?%s?=\"" % [attr, 'ISO-2022-JP', 'B', str.encode_cs('ISO-2022-JP', 'UTF-8').encode_ec('B').chomp])
}

#     漢字: 0xA1-0xFE 0xA1-0xFE
# 半角カナ: 0x8E 0xA1-0xDF	
# 補助漢字: 0x8F 0xA1-0xFE 0xA1-0xFE

String.bind_wsizer('EUC-JP') {|str|								# 表示幅を得る
	str.force_encoding('ASCII-8BIT').gsub(/[\xA1-\xFE][\xA1-\xFE]/n, "\xFF\xFF").length
}

String.bind_centerer('EUC-JP') {|str, n, padding|				# センタリングする
	w = n - str.wsize
	w = 0 if(w < 0)
	(padding * (w >> 1) + str + padding * w).snip(n)
}

String.bind_snipper('EUC-JP') {|str, n|							#### 指定の長さに切り詰める
	str.force_encoding('ASCII-8BIT')
	ws = str[0, n * 2].gsub(/[\xA1-\xFE][\xA1-\xFE]/n, String::XFFFF)[0, n]
	wc = ws.count(String::XFF)
	str.slice(0, n - wc % 2) + ' ' * (n - ws.length + wc % 2)
}

String.bind_each_snipper('EUC-JP') {|str, n, max, proc|			#### 指定の長さに切り詰め、順に行を渡す
	str.force_encoding('ASCII-8BIT')
	p = 0; while(p <= str.length)								# '<': 改行文字のみの行は省略
		break if((max -= 1) < 0)
		ws = str[p, n * 2].gsub(/[\xA1-\xFE][\xA1-\xFE]/n, String::XFFFF)[0, n]
		wc = ws.count(String::XFF)
		proc.call(str.slice(p, nn = n - wc % 2) + ' ' * (n - ws.length + wc % 2))
		p += nn
	end
}

#### TAB 対応

__END__

